contracts/contracts/Erdstall.sol

```diff
@@ -101,63 +99,99 @@ contract Erdstall {
     // Challenge Functions
     //
 
-    // Challenges the operator to post the epoch's balance statement.
+    // Challenges the operator to post the current exit epoch's balance statement.
+    // The user needs to pass the latest balance proof, that is, of the just
+    // sealed epoch, to proof that they are part of the system.
+    //
+    // After a challenge is opened, the operator (anyone, actually) can respond
+    // to the challenge using function `exit`.
+    function challenge(Balance calldata balance, bytes calldata sig) external onlyAlive {
+        require(balance.account == msg.sender, "challenge: wrong sender");
+        require(balance.epoch == sealedEpoch(), "challenge: wrong epoch");
+        verifyBalance(balance, sig);
+
+        registerChallenge(balance.value);
+    }
+
+    // challengeDeposit should be called by a user if they deposited but never
+    // received a deposit or balance proof from the operator.
+    //
     // After a challenge is opened, the operator (anyone, actually) can respond
-    // to the challenge using function exit.
-    function challenge() external onlyAlive {
+    // to the challenge using function `exit`.
+    function challengeDeposit() external onlyAlive {
+        registerChallenge(0);
+    }
+
+    function registerChallenge(uint256 recoveryBalance) internal {
         require(!isChallengeResponsePhase(), "in challenge response phase");
         uint64 epoch = exitEpoch();
-        require(!challenges[epoch][msg.sender], "already challenged");
+        require(challenges[epoch][msg.sender] == 0, "already challenged");
+
+        uint256 value = recoveryBalance + deposits[epoch][msg.sender];
+        require(value > 0, "no value in system");
 
-        challenges[epoch][msg.sender] = true;
+        challenges[epoch][msg.sender] = value;
         numChallenges[epoch]++;
 
         emit Challenged(epoch, msg.sender);
     }
 
-    // Freezes the contract such that only exits of the prior exit phase can be
-    // posted.
-    // Can be called by any user if the prior phase has an open challenge.
-    // This must be called in the phase following the challenged exit phase.
-    function freeze() external {
```

Comment:
What do you think about keeping a method `freeze` for convenience and clarity? It's essential that `withdrawChallenge` is called in the immediately following epoch after the challenged epoch, right? With withdrawals you could usually also wait until later periods, right? With the current naming, somebody could think he can also just withdraw later, but then it's too late.

Reply:
Already there, now called `ensureFrozen public`. Note that `public` is both, `internal` and `external`.